### ISSUE TITLE
Logic for triggering Behat tests

### ISSUE THREAD

**Issue by stevector** [pantheon-systems/example-wordpress-composer#40]
Follow up from discussion in https://github.com/pantheon-systems/example-wordpress-composer/pull/34#issuecomment-328122556

The Behat script checks to see if the commit being tested is in a pull request or on master before proceeding with the test. We've found problematic false passes and false fails.

1. False fail: The Behat test runs because a pull request is present. But the separate job running earlier in the workflow to create the Multidev being tested ran before the pull request was created. And thus no Multidev is created. Behat then fails because there is no environment to run against.
2. False pass. It seems the `$CIRCLE_PULL_REQUEST` variable isn't always populated when it should be. Here is a build from after a PR was made where the CIRCLE_PULL_REQUEST is missing. https://circleci.com/gh/pantheon-systems/example-wordpress-composer/475?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link The build then exits with no error and no Behat test actually running.

**Comment by greg-1-anderson** [comment]
I think that rather than complicating things too much, we should simply:

- Create a multidev called `pr-PULL_REQUEST_NUMBER` when there is a PR available.
- Create a multidev called `ci-CIRCLE_BUILD_NUMBER` when there is no PR available.

The later is already done for builds on the `master` branch, for example. It would be a simple enough thing to also do this for the first build on a branch that will (or may) become a pull request later. These multidevs can be cleaned up per the existing policy (keep three around, delete the rest).

I am already doing this in my [work-in-progress PR to convert drops-8 to Circle 2.0](https://github.com/pantheon-systems/example-drops-8-composer/pull/107/files#diff-a2a81420ddea50471f5fb05dcb4ff8bcR21).

**Comment by stevector** [comment]
Part of @ataylorme reasoning for checking for PRs at all was to avoid time consuming Behat test runs if there is not a PR. Instead, we'd get faster/pass fail feedback on non-PRs.

I think that concern is mitigated by CircleCI 2 reporting on separate jobs. The feedback on whether code sniffing passes comes very fast. I'd like to simplify and take @greg-1-anderson's suggestion here:

> Create a multidev called pr-PULL_REQUEST_NUMBER when there is a PR available.
> Create a multidev called ci-CIRCLE_BUILD_NUMBER when there is no PR available.

And run Behat tests by default every time. We can still document how to bail on Behat for non-PRs if there is a team out there that needs to do that.

**Comment by stevector** [comment]
That being said, if the we are keeping separate the jobs that create the multidev and the job that runs the Behat test (and I think we are) then those jobs might get different answers to the question "Is there a PR?". So I think we do need some way of passing along the name of the created multidev from `build_and_deploy` to `behat_test`.

**Comment by ataylorme** [comment]
> Part of @ataylorme reasoning for checking for PRs at all was to avoid time consuming Behat test runs if there is not a PR

Prior to a pull request I strongly feel it's unnecessary to spin up a multidev on Pantheon at all.

The main reasoning here is that when a branch is pushed we should do some low fidelity tests, like code sniffing. At that point since no pull request is open the work is not ready for review.

Once a pull request is open, and the work is ready for review, we want to run more exhaustive tests, such as Behat, in addition to a code review on GitHub.

I think two solutions to achieve this are:
1) Have the `build_and_deploy` job trigger the `behat_test` job via API if a pull request is open
2) Check for a pull request on the first job/script that runs, set our own variable and cache it for use in the subsequent scripts/jobs. This would essentially be taking the variables in `set-up-globals.sh` that are exported to `$BASH_ENV` and sourced at the start of each job and defining them once and writing them to a file to be sourced rather than redefining them each time.

I would prefer the second option.

**Comment by greg-1-anderson** [comment]
I felt that the number of distinct steps in the circle 2.0 tests were adding complexity that was not needed. In the drops-8 example, I simplified to just two:

- `build` does the code sniffing
- `build_deploy_and_test` runs behat

There is no need to trigger the behat test via the API; you can just wrap the test in an `if` statement if you do not want it to run.

See https://github.com/pantheon-systems/example-drops-8-composer/pull/107 for the latest version of my tests. A couple of scripts still need to be generalized before these could be pulled back to the WordPress project.

**Comment by ataylorme** [comment]
@greg-1-anderson two jobs makes sense to me. I would, however, still argue that `build_deploy_and_test` isn't needed if there isn't a pull request.

**Comment by greg-1-anderson** [comment]
We do need to run the `build_deploy_and_test` step for the master branch, and any feature branch that is being used like master (e.g. a `1.x` branch, et. al.).  In the previous iteration of these tests I distinguished between these branches and branches-that-might-be-pull-requests. At this point in time, I do not think it is worth adding too much complexity to avoid the tests when there is no pull request. If it could be done simply, I'm in favor (or at a minimum indifferent).

**Comment by ataylorme** [comment]
I guess the disadvantages I see to always running `build_deploy_and_test` are:
* having the `ci-` multidevs (which have had naming conflicts in the past)
* If a developer hasn't opened a pull request they are more likely to push updates in rapid succession. This is a larger problem of CircleCI race conditions and running multiple sets of the CircleCI workflow at once (e.g. two builds competing to spin up a multidev or run Behat tests)
* when a pull request is opened we'll also have a `pr-` multidev in addition to the `ci-` multidev, though this will be cleaned up eventually
* the time in which it takes CircleCI to report back to GitHub on success/failure is greatly increased when not making a pull request as multidev creation and behat tests are time consuming

I am okay with these things though and, stealing your wording, in favor of (or at a minimum indifferent) to moving forward as-is.

**Comment by greg-1-anderson** [comment]
The `ci-` multidevs should not have naming conflicts, as they are named after the Circle build number. You could get naming conflicts if you use one Pantheon site to test multiple GitHub repositories; this is not recommended, though; this would also cause conflicts with the `pr-` multidevs.

I haven't really seen Circle get into a race condition with multiple builds; sometimes, some builds stall out while older builds finish.

CircleCI shouldn't report success/failure back to GitHub for non-pull requests.

I've sort of gotten used to making a lot of commits, but not necessarily pushing them until I am ready to make a PR. If devs are not in this habit, then I can see the value of skipping the time-consuming tests. Hopefully folks will at a minimum avoid pushing rapidly to the `master` branch, as we will want to run the test every time there.

**Comment by greg-1-anderson** [comment]
As @ataylorme [mentions in another issue](https://github.com/pantheon-systems/example-wordpress-composer/issues/28), there is a potential problem with using the Circle build number to name the multidev used for the site-under-test, as every build step gets its own build number.

**Comment by greg-1-anderson** [comment]
There is also a workflow ID:
```
CIRCLE_WORKFLOW_ID=b935e0b1-6ca9-4577-82ad-77016922c8d4
```
The problem here is obvious: the ID itself is far too long to use in a multidev name, and if we reduced it, then there is a chance for collision. It also becomes more difficult to order `ci-*` environments for deletion.

**Comment by ataylorme** [comment]
I wish CircleCI's API allowed us to get information about workflows. We could then use the earliest job number in that workflow, a timestamp of when the workflow started, etc.

**Comment by greg-1-anderson** [comment]
I made a support request asking for exactly that.

Text from my support request:

> 
> In Circle 1.x, we would name external resources needed for test fixtures based on the CIRCLEBUILDNUM.
> 
> In Circle 2.0, every build step gets a different CIRCLEBUILDNUM, making it difficult for follow-on steps to find the external test fixtures created in an earlier step.
> We do have CIRCLEWORKFLOWID at our disposal; however, this value is too long for us to use in our fixture names. If we hash this value down, there is a chance for collisions. The other problem with CIRCLEWORKFLOWID is that, hashed down or not, these values are not sequential. We used the fact that CIRCLEBUILDNUM is monotonically increasing to help us lazy-delete our old test fixtures.
> 
> We do not want to use CIRCLE_BRANCH et. al., as we would like a value that differs for different workflows ran against the same branch.
> 
> Proposal:
> 
> Please define a new environment variable CIRCLEWORKFLOWNUMBER. This should be a monotonically increasing integer value. It could have the same value as the CIRCLEBUILDNUM for the first step in the workflow that is assigned a number, or it could be independent from the build number.

**Comment by greg-1-anderson** [comment]
Circle support pointed me at an [existing Circle CI discussion](https://discuss.circleci.com/t/2-0-need-circle-workflow-id-env-var-exposed/14200/11) about this feature request.  Those who want this feature should LIKE comments in that discussion, as Circle CI apparently evaluates community group discussion LIKES when prioritizing future work.
